util/pyclient/bft_client.py

```diff
@@ -241,3 +247,256 @@ def get_rsi_replies(self):
+                # Wrap this stream with SSL stream, pass server_hostname to be verified
+                ssl_stream = trio.SSLStream(tcp_stream, ssl_context, server_hostname=server_hostname, https_compatible=False)
+                # Wait for handshake to finish (we want to be on the safe side - after this we are sure 
+                # connection is open)
+                await ssl_stream.do_handshake()
+                # Success! keep stream in dictionary and break out
+                self.ssl_streams[(dest_replica.ip, dest_replica.port)] = ssl_stream
+                return
+            except (OSError, trio.BrokenResourceError):
+                await trio.sleep(0.25)
+                pass
+            continue
+
+    async def on_stopped_replicas(self, replica_ids):
+        """ BFT network inform client when a group of replicas are stopped. We relay on caller cancel scope """
+        for id in replica_ids:
+            replica = self.replicas[id]
+            addr = (replica.ip, replica.port)
+            if addr in self.ssl_streams:
+                if addr in self.ssl_streams:
+                    stream = self.ssl_streams[addr]
+                    del self.ssl_streams[addr]
+                    await stream.aclose()
+
+    async def _send_data(self, data, dest_replica):
+        """ Try to send data to dest_replica. On exception - close the SSL stream. """
+        dest_addr = (dest_replica.ip, dest_replica.port)
+        if dest_addr not in self.ssl_streams.keys():
+            # dest_replica is not connected (crushed? isolated?) 
+            return
+        # first 4 bytes include the data header (message size), then comes the data
+        data_len = len(data)
+        out_buff = bytearray(data_len.to_bytes(self.MSG_HEADER_SIZE, "little"))
+        out_buff += bytearray(data)
+        stream = self.ssl_streams[dest_addr]
+        try:
+            await stream.send_all(out_buff)
+            return True
+        except (trio.BrokenResourceError, trio.ClosedResourceError):
+            # Failed! close the stream and return failure.
+            if dest_addr in self.ssl_streams:
+                del self.ssl_streams[dest_addr]
+                await stream.aclose()
+            return False
+
+    async def _stream_recv_some(self, out_data, dest_addr, stream, num_bytes):
+        """ Try to receive data from replica. On failure - close the SSL stream. """
+        try:
+            out_data += await stream.receive_some(num_bytes)
+            return True
+        except (trio.BrokenResourceError, trio.ClosedResourceError):
+            if dest_addr in self.ssl_streams:
+                del self.ssl_streams[dest_addr]
+                await stream.aclose()
+            return False
+
+    async def _receive_from_replica(self, dest_addr, replicas_addr, required_replies, cancel_scope):
+        """
+        Receive from a single replica. 3 stages:
+        1) Wait for the 1st 4 bytes which consist the header (payload length)
```

Comment:
nit - "... which contain the header ..."

Reply:
"which contains the header" 👍